include/mbedtls/lms.h

```diff
@@ -0,0 +1,254 @@
+ *                           this.
+ *
+ * \return         \c 0 on success.
+ * \return         A non-zero error code on failure.
+ */
+int mbedtls_lms_import_pubkey( mbedtls_lms_context *ctx,
+                               const unsigned char *key );
+
+/**
+ * \brief                    This function exports an LMOTS public key from a
+ *                           LMS context that already contains a public key.
+ *
+ * \note                     Before this function is called, the context must
+ *                           have been initialized and the context must contain
+ *                           a public key.
+ *
+ * \note                     See IETF RFC8554 for details of the encoding of
+ *                           this public key.
+ *
+ * \param ctx                The initialized LMS context that contains the
+ *                           publc key.
+ * \param key                The buffer into which the key will be output. Must
+ *                           be at least #MBEDTLS_LMS_PUBKEY_LEN in size.
+ *
+ * \return         \c 0 on success.
+ * \return         A non-zero error code on failure.
+ */
+int mbedtls_lms_export_pubkey( mbedtls_lms_context *ctx,
+                               unsigned char *key );
+
+/**
+ * \brief                    This function generates an LMS public key from a
+ *                           LMS context that already contains a private key.
+ *
+ * \note                     Before this function is called, the context must
+ *                           have been initialized and the context must contain
+ *                           a private key.
+ *
+ * \param ctx                The initialized LMS context to generate the key
+ *                           from and store it into.
+ *
+ * \return         \c 0 on success.
+ * \return         A non-zero error code on failure.
+ */
+int mbedtls_lms_gen_pubkey( mbedtls_lms_context *ctx );
+
+/**
+ * \brief                    This function generates an LMS private key, and
+ *                           stores in into an LMS context.
+ *
+ * \note                     Before this function is called, the context must
+ *                           have been initialized and the type of the LMS
+ *                           context set using mbedtls_lmots_set_algorithm_type
+ *
+ * \note                     The seed must have at least 256 bits of entropy.
+ *
+ * \param ctx                The initialized LMOTS context to generate the key
+ *                           into.
+ * \param f_rng              The RNG function to be used to generate the key ID.
+ * \param p_rng              The RNG context to be passed to f_rng
+ * \param seed               The seed used to deterministically generate the
+ *                           key.
+ * \param seed_len           The length of the seed.
+ *
+ * \return         \c 0 on success.
+ * \return         A non-zero error code on failure.
+ */
+int mbedtls_lms_gen_privkey( mbedtls_lms_context *ctx,
```

Comment:
There are a couple of tricky things with storing the private key, notably:
1) The RFC states that the private key format is implementation defined, so we need to create a representation (and it won't work anywhere else)
2) Storing the private key means also being able to load it into a context, which then leads to an issue where the same   private key can be imported into multiple contexts simultaneously, when would then lead to subkey reuse. This _can_ happen currently if someone copies an `mbedtls_lms_context`, and we definitely need a docs warning about that, but it would seem to be a less likely thing to happen than if we had an API to import and export keys
3) We'd end up exposed to a lot of complexity regarding updating `q_next_usable_key`, as you point out. Currently, we guarantee that the next key index is updated in the context before the signature is emitted, so we're relatively safe there, though of course this isn't much help when you need to store the private key in non-volatile storage. Other implementations (https://github.com/cisco/hash-sigs/blob/master/hss.h#L195) seem to solve this by adding a function pointer to the signature which is used to update the key in non-volatile storage. This adds a lot of complexity that the caller has to handle, but also does seem to be a workable solution

Reply:
I've had time to have a think about import/export now. Effectively our problem is that any scenario where you can copy a stored key is dangerous, because that leaves you open to reusing subkeys. The real issue is that we're a security library, and the device code _absolutely_ has the power to copy arbitrary memory, so we can't solve this issue completely with technology. There are some things we can do easily to discourage copying keys, but we're effectively looking for a sensible midpoint where on the spectrum between making it impossible and having a policy which says "don't copy keys".

I propose:
1) Adding a function pointer to the arguments of `mbedtls_lms_sign`, with the signature `int save_key(size_t len, uint8_t *key)`, which at least makes it possible to update keys on-disk / in flash storage. Note that once they're there we _absolutely_ cannot stop people copying them
2) Adding a private key loading function of type `int mbedtls_lms_import_privkey(mbedtls_lms_context *ctx, uint8_t* key)`. 
 * This function is tricky since we can't possibly provide loaders for every possible place a key might be stored. The tricky one is storage locations that are not memory-mapped, since in that case the caller would need to first load the key into RAM and then call import on that, and there's potential for issues there (since the callback function above would likely update the on-disk key) where we have two copies of the same key that are out of sync.
  * One potential option would be again function pointers - we could have the function be called solely with a function pointer of type `int load_key(size_t len, uint8_t *key)`, which we'd the supply the buffer to and call. This means that the transient copy is under our control (or ideally is avoided altogether), but this is a pretty weird pattern for the API to have.
3) Changing the `q_next_usable_key` field of the lms context to be an `unsigned int *`, then dynamically allocating it in a `_setup` function (which would be the renamed `set_algorithm_type`. This means that any trivial copying of an lms context will mean they share the `q_next_usable_key` pointer, which is safe unless in a threaded environment.
4) Potentially changing the `q_next_usable_key` pointer to be a pointer to an opaque struct (defined in lms.c). This should prevent manually reallocating the `q_next_usable_key` pointer, since the library user won't know the size (unless they do a best-guess memcpy)
5) Writing some documentation to make clear that having multiple copies of a single stateful key is a security risk.

Reply:
Save/load, which is necessary in the real world, is even more of a problem, because of resiliency. There are many realistic scenarios where an adversary can cause a system crash or power failure. So, if it's important that the state of an object is never rolled back, just saving the updated state after updating it is not enough:

* 1. Consume subkey in RAM.
* Power failure (accidental or adversarial).
* ~2. Save updated state. ~

Instead the API must allow this scenario, and should try to prevent other scenarios:

* 1. Save updated state with one more subkey consume.
* 2. Use up the updated subkey.

That way a power failure at the wrong time can lead to resource exhaustion but not to key exposure.

Reply:
This is one of the core parts of dealing with the stateful keys, and is solved (though the exhaustion is still an issue) by updating the `q_next_usable_key` value before even computing the signature. We already do this for in-ram keys (see `lms.c:378`), and I think adding the function pointer to save the key (which would be called at the same point) would solve this problem for keys stored in non-RAM mediums.

As you point out, exhaustion is still an issue, but I don't see a way to solve this while keeping the safety of the key updating. We may need some docs that mentions this issue and makes users aware that they should consider it when sizing keys.

Reply:
So I think for now the idea is to relegate the LMS signature functionality to testing purposes only, because there are some pretty tricky issues related to it. I've added a note to the function that it's for testing purposes only, but I'm not sure if that's enough. Do you think something stronger is required?